### PR TITLE
feat: support multi-client for ckb-eth

### LINKED ISSUE #223: Support multi-client ETH-CKB relay
TODO list
- [ ] `minimal_updates_count`. We currently only support `minimal_updates_count = 1`:
    1. When creating new multi-client cells, only one update is given .
    2. There is no guarantee that updates will contains at least `minimal_updates_count` 
- [ ] How to record new client's type_id?
    1. Writing it back to `config.toml` is an option, but it carries some risk as changing the config file during runtime may be risky.
    2. Write it to a record in `data_dir`.
    3. Users can opt to manually record it, as it only needs to be done once.
- [x] Use official eth-light-client-in-ckb repo
- [ ] Lock script. 
- [ ] Fix tests.

### PR DESCRIPTION
**Pull request by whfuyn**
Closes: #XXX

## Description

______

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests: integration (for Forcerelay) or unit/mock tests (for modules).
- [ ] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: crates/relayer-storage/Cargo.toml [hunk 1]
```diff
@@ -15,4 +15,5 @@ description  = "The storage part of SynapseWeb3 IBC Relayer"
 thiserror = "1.0.37"
 rocksdb = { package = "ckb-rocksdb", version ="=0.19.0", default-features = false, features = ["snappy"] }
 eth2_types = { git = "https://github.com/synapseweb3/lighthouse", rev = "2c246d6", package = "types" }
-eth_light_client_in_ckb-verification = { git = "https://github.com/synapseweb3/eth-light-client-in-ckb", branch = "release/v0.1.x" }
+# eth_light_client_in_ckb-verification = { git = "https://github.com/yangby-cryptape/eth-light-client-in-ckb", rev = "5a5dc2ee18736d492d0a73de691a375f0daf9ff3" }
+eth_light_client_in_ckb-verification = { version = "0.2.0-alpha", git = "https://github.com/whfuyn/eth-light-client-in-ckb", rev = "6c2760343a42cd1f6074cd11620291da9467b60e" }
```
Comment (ashuralyk) [inline crates/relayer-storage/Cargo.toml]: what have you modified in this boyu's repo?

File: crates/relayer/src/chain/ckb.rs [hunk 2]
```diff
@@ -156,63 +195,148 @@ impl CkbChain {
...
+            )?
+        else {
+            return Err(Error::other_error("no multi-client cells found".to_owned()));
+        };
+
+        let latest_client = PackedClient::new_unchecked(update_cells.latest.output_data.clone());
+        self.cached_onchain_packed_client = Some(latest_client);
+
+        let minimal_updates_count = {
+            let client_info = PackedClinetInfo::new_unchecked(update_cells.info.output_data.clone());
```
Comment (ashuralyk) [inline crates/relayer/src/chain/ckb.rs]: `PackedClinetInfo` should be `PackcedClientInfo`

File: crates/relayer/src/chain/ckb.rs [hunk 3]
```diff
@@ -156,63 +195,148 @@ impl CkbChain {
...
         self.sign_and_send_transaction(tx, inputs).map_err(|err| {
             if let Err(err) = self.storage.rollback_to(prev_slot_opt) {
                 return err.into();
             }
             err
         })?;
 
+        // TODO: Write back the type id to config.
+        tracing::info!("new type_id: {}", type_id);
+        self.config.client_type_args.type_id = Some(type_id);
```
Comment (ashuralyk) [inline crates/relayer/src/chain/ckb.rs:208]: how to persist this config object?

**Review by ashuralyk — COMMENTED** [review]
this tough feature is accomplished almost entirely, although I think it still has posibility to wrap some code logic into utilities

File: crates/relayer-storage/Cargo.toml (hunk 1 above)
Comment (whfuyn) [reply crates/relayer-storage/Cargo.toml]: At the time I change this dep url to my fork, both [this](https://github.com/synapseweb3/eth-light-client-in-ckb/pull/7) and [this](https://github.com/synapseweb3/eth-light-client-in-ckb/pull/6) are not merged. So I made my fork to include those changes.

We can change it back to official repo when it's merged.

File: crates/relayer/src/chain/ckb.rs (hunk 3 above)
Comment (whfuyn) [reply crates/relayer/src/chain/ckb.rs:208]: As mentioned in #223 , I'm considering three ways to record this `type_id`:
      1. Writing it back to `config.toml` is an option, but it carries some risk as changing the config file during runtime may be risky.
      2. Write it to a record in `data_dir`.
      3. Users can opt to manually record it, as it only needs to be done once.

Comment (ashuralyk) [reply crates/relayer/src/chain/ckb.rs:208]: temporarily we can persist it manually in `config.toml` at this moment for examing the feasibility of the multi-relayer feature, it won't block our work

**Comment by ashuralyk** [comment]
suggest to resolve conflicts

**Review by ashuralyk — APPROVED** [review]
after merging it, please put a tag on this to show this is a multi-client version which is the phase 2 for Forcerelay/Eth

**Pull request merged**